crates/nu-cli/src/shell/filesystem_shell.rs

```diff
@@ -1000,125 +1002,143 @@ impl Shell for FilesystemShell {
-                                                    yield Ok(ReturnSuccess::Value(
-                                                        val,
-                                                    ))
-                                                },
-                                            }
-                                        }
-                                    } else {
-                                        if f.is_dir() {
-                                            let msg = format!(
-                                                "Cannot remove {:}. try --recursive",
-                                                f.to_string_lossy()
-                                            );
-                                            yield Err(ShellError::labeled_error(
-                                                msg,
-                                                "cannot remove non-empty directory",
-                                                &target.tag,
-                                            ))
-                                        } else {
-                                            let msg = format!("Invalid file: {:}", f.to_string_lossy());
-                                            yield Err(ShellError::labeled_error(
-                                                msg,
-                                                "invalid file",
-                                                &target.tag,
-                                            ))
-                                        }
-                                    }
+            let mut path = PathBuf::from(path);
+            path.push(&target.item);
+            match glob::glob(&path.to_string_lossy()) {
+                Ok(files) => {
+                    for file in files {
+                        match file {
+                            Ok(ref f) => {
+                                if !unique_paths.contains(f) {
+                                    unique_paths.insert(f.clone());
+                                    all_targets.push(Tagged::new(target.tag.clone(), f.clone()));
                                 }
+                            }
+                            Err(e) => {
+                                let msg = format!("Could not remove {:}", path.to_string_lossy());
+                                return Err(ShellError::labeled_error(
+                                    msg,
+                                    e.to_string(),
+                                    &target.tag,
+                                ));
+                            }
+                        }
+                    }
+                }
+                Err(e) => {
+                    return Err(ShellError::labeled_error(
+                        format!("Remove aborted. {:}", e.to_string()),
+                        e.to_string(),
+                        &name_tag,
+                    ))
+                }
+            };
+        }
```

Comment:
nit: if you'd like, you could simplify this a lot by just using a `HashSet` for `all_targets`, and [`HashSet.extend`](https://doc.rust-lang.org/std/collections/struct.HashSet.html#impl-Extend%3CT%3E).

The use of `all_targets` below should just work with a `HashSet`, since you can get an `iter` from `HashSet` too 🙂

Reply:
The reason I included the HashSet is to detect targets that are specified in more than one path/glob. Trying to delete these twice caused an error. 

I agree that this is not a particularly clean solution. As you suggested, I tried using only the HashSet, but because I need to store the tag from the original path/glob with each target, this didn't work. The HashSet seemed to allow two identical paths if they were tagged differently. Do you know of a clean way around this?

Another reason I liked the vector is that if deletion aborts due to the second path argument, then the first will still have been deleted, but any remaining arguments won't have been. Maintaining this ordering made more sense to me than adopting some ordering based on the HashSet.

Reply:
Ah, that makes total sense, but

> The HashSet seemed to allow two identical paths if they were tagged differently.

There's another option! 🙂

How about using a `HashMap<PathBuf, Tag>`, that way the  uniqueness comparison is on the `PathBuf`, but you can still maintain the tag information for it? You may have to make some small tweaks below, since an `iter` will look slightly different for a `HashMap`, but I think it'll simplify this a bunch. What do you think?

Reply:
That's a great idea, thank you!